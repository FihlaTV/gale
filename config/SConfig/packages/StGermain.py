import os
import SConfig

class StGermain(SConfig.Package):
    def __init__(self, env, options):
        SConfig.Package.__init__(self, env, options)
        self.require(SConfig.packages.cmath)
        self.require(SConfig.packages.libXML2)
        self.require(SConfig.packages.MPI)
        self.base_patterns = ['StGermain*']
        self.headers = [[os.path.join('StGermain', 'StGermain.h')]]
        self.libraries = [['StGermain']]
        self.symbols = [(['StGermain_Init', 'StGermain_Finalise'], '')]
        self.symbol_setup = 'MPI_Init(&argc, &argv);'
        self.symbol_teardown = 'MPI_Finalize();'
        self.symbol_calls = ['%s(&argc, &argv);', '%s();']
        self.require_shared = True
        self.use_rpath = True

    def get_run_error_message(self, console):
        if len(console):
            return 'Incompatible libraries, check \'config.log\'.'
