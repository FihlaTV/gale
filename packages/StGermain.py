import os
import config

class StGermain(config.Package):

    def __init__(self, ctx, **kw):
        config.Package.__init__(self, ctx, **kw)
        self.inc_exts = ["StGermain"]

    def setup_dependencies(self):
        config.Package.setup_dependencies(self)
        self.add_dependency(config.packages.MPI, required=True, combine=True)
        self.add_dependency(config.packages.libxml2, required=True, combine=True)

    def setup_libraries(self):
        self.add_library_set([os.path.join("StGermain", "StGermain.h")], ["StGermain"])

    def setup_options(self):
        config.Package.setup_options(self)
        self.options.add_option(
            config.utils.options.Option("stg_dir", "--stg-dir",
                                        pref_sep="=", help=self.help["stg_dir"]))

    def process_options(self):
        config.Package.process_options(self)
        base_dir = self.get_option("stg_dir", None)
        if base_dir is not None:
            self.forced_base_dirs = [base_dir]

    def make_help_dict(self):
        config.Package.make_help_dict(self)
        self.help["stg_dir"] = "Specify the base location for all StG* related projects."

    source_code = {"c": """#include <stdlib.h>
#include <StGermain/StGermain.h>
int main( int argc, char** argv ) {
  return EXIT_SUCCESS;
}
"""}
