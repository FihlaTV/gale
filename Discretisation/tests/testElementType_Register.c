/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003-2006, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street,
**	Melbourne, 3053, Australia.
**
** Primary Contributing Organisations:
**	Victorian Partnership for Advanced Computing Ltd, Computational Software Development - http://csd.vpac.org
**	Australian Computational Earth Systems Simulator - http://www.access.edu.au
**	Monash Cluster Computing - http://www.mcc.monash.edu.au
**	Computational Infrastructure for Geodynamics - http://www.geodynamics.org
**
** Contributors:
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Robert Turnbull, Research Assistant, Monash University. (robert.turnbull@sci.monash.edu.au)
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	David May, PhD Student, Monash University (david.may@sci.monash.edu.au)
**	Louis Moresi, Associate Professor, Monash University. (louis.moresi@sci.monash.edu.au)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**	Julian Giordani, Research Assistant, Monash University. (julian.giordani@sci.monash.edu.au)
**	Vincent Lemiale, Postdoctoral Fellow, Monash University. (vincent.lemiale@sci.monash.edu.au)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
** $Id: testElementType_Register.c 832 2007-05-16 01:11:18Z LukeHodkinson $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include <StGermain/StGermain.h>
#include "StgFEM/Discretisation/Discretisation.h"

#include <stdio.h>
#include <stdlib.h>
#include <petsc.h>

int main( int argc, char* argv[] ) {
	MPI_Comm			CommWorld;
	int				rank;
	int				numProcessors;
	int				procToWatch;
	ElementType_Register*		elementType_Register;
	Stream*				stream;
	
	/* Initialise MPI, get world info */
	MPI_Init( &argc, &argv );
	MPI_Comm_dup( MPI_COMM_WORLD, &CommWorld );
	MPI_Comm_size( CommWorld, &numProcessors );
	MPI_Comm_rank( CommWorld, &rank );
	
	StGermain_Init( &argc, &argv );
	StgFEM_Discretisation_Init( &argc, &argv );
	MPI_Barrier( CommWorld ); /* Ensures copyright info always come first in output */
	
	stream = Journal_Register (Info_Type, "myStream");

	if( argc >= 2 ) {
		procToWatch = atoi( argv[1] );
	}
	else {
		procToWatch = 0;
	}
	if( rank == procToWatch ) printf( "Watching rank: %i\n", rank );
	
	/* Build the element type register */
	elementType_Register = ElementType_Register_New("elementTypeRegister");
	
	/* Add the prebuilt types */
	ElementType_Register_Add( elementType_Register, (ElementType*)ConstantElementType_New("constant") );
	ElementType_Register_Add( elementType_Register, (ElementType*)BilinearElementType_New("bilinear") );
	ElementType_Register_Add( elementType_Register, (ElementType*)TrilinearElementType_New("trilinear") );
	
	/* Manually create extra types to force test the list re-sizing */
	ElementType_Register_Add( elementType_Register, _ElementType_New( sizeof(ConstantElementType), "TestElementType0",
		_ConstantElementType_Delete, _ConstantElementType_Print, NULL, ConstantElementType_DefaultNew,
		_ConstantElementType_Construct, _ConstantElementType_Build, _ConstantElementType_Initialise,
		_ConstantElementType_Execute, _ConstantElementType_Destroy, "TestElementType0Name", True,
		_ConstantElementType_SF_allNodes, _ConstantElementType_SF_allLocalDerivs_allNodes,
		_ConstantElementType_ConvertGlobalCoordToElLocal,
		1 ) );
	
	ElementType_Register_Add( elementType_Register, _ElementType_New( sizeof(ConstantElementType), "TestElementType1",
		_ConstantElementType_Delete, _ConstantElementType_Print, NULL, ConstantElementType_DefaultNew,
		_ConstantElementType_Construct, _ConstantElementType_Build, _ConstantElementType_Initialise,
		_ConstantElementType_Execute, _ConstantElementType_Destroy, "TestElementType1Name", True,
		_ConstantElementType_SF_allNodes, _ConstantElementType_SF_allLocalDerivs_allNodes,
		_ConstantElementType_ConvertGlobalCoordToElLocal,
		1 ) );
	
	ElementType_Register_Add( elementType_Register, _ElementType_New( sizeof(ConstantElementType), "TestElementType2",
		_ConstantElementType_Delete, _ConstantElementType_Print, NULL, ConstantElementType_DefaultNew,
		_ConstantElementType_Construct, _ConstantElementType_Build, _ConstantElementType_Initialise,
		_ConstantElementType_Execute, _ConstantElementType_Destroy, "TestElementType2Name", True,
		_ConstantElementType_SF_allNodes, _ConstantElementType_SF_allLocalDerivs_allNodes,
		_ConstantElementType_ConvertGlobalCoordToElLocal,
		1 ) );
	
	ElementType_Register_Add( elementType_Register, _ElementType_New( sizeof(ConstantElementType), "TestElementType3",
		_ConstantElementType_Delete, _ConstantElementType_Print, NULL, ConstantElementType_DefaultNew,
		_ConstantElementType_Construct, _ConstantElementType_Build, _ConstantElementType_Initialise,
		_ConstantElementType_Execute, _ConstantElementType_Destroy, "TestElementType3Name", True,
		_ConstantElementType_SF_allNodes, _ConstantElementType_SF_allLocalDerivs_allNodes,
		_ConstantElementType_ConvertGlobalCoordToElLocal,
		1 ) );
	
	ElementType_Register_Add( elementType_Register, _ElementType_New( sizeof(ConstantElementType), "TestElementType4",
		_ConstantElementType_Delete, _ConstantElementType_Print, NULL, ConstantElementType_DefaultNew,
		_ConstantElementType_Construct, _ConstantElementType_Build, _ConstantElementType_Initialise,
		_ConstantElementType_Execute, _ConstantElementType_Destroy, "TestElementType4Name", True,
		_ConstantElementType_SF_allNodes, _ConstantElementType_SF_allLocalDerivs_allNodes,
		_ConstantElementType_ConvertGlobalCoordToElLocal,
		1 ) );
	
	ElementType_Register_Add( elementType_Register, _ElementType_New( sizeof(ConstantElementType), "TestElementType5",
		_ConstantElementType_Delete, _ConstantElementType_Print, NULL, ConstantElementType_DefaultNew,
		_ConstantElementType_Construct, _ConstantElementType_Build, _ConstantElementType_Initialise,
		_ConstantElementType_Execute, _ConstantElementType_Destroy, "TestElementType5Name", True,
		_ConstantElementType_SF_allNodes, _ConstantElementType_SF_allLocalDerivs_allNodes,
		_ConstantElementType_ConvertGlobalCoordToElLocal,
		1 ) );
	
	
	Stg_Class_Print( elementType_Register, stream );
	
	/* Destroy stuff */
	Stg_Class_Delete( elementType_Register );
	
	StgFEM_Discretisation_Finalise();
	StGermain_Finalise();
	
	/* Close off MPI */
	MPI_Finalize();
	
	return 0; /* success */
}

