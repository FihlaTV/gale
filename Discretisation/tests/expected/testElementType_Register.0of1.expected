StGermain Framework. Copyright (C) 2003-2005 VPAC.
StGermain Discretisation Library. Copyright (C) 2003-2005 VPAC.
StGermain FEM Discretisation Framework. Copyright (C) 2003-2005 VPAC.
Watching rank: 0
ElementType_Register (ptr): 0x8064000
Stg_Class (ptr): 0x8064000
	sizeOfSelf: 44
	_deleteSelf: Yes
	type: ElementType_Register
	_delete (func ptr): 0x804b9f5
	_print (func ptr): 0x804ba8b
	_copy (func ptr): (nil)
	count: 9
	_size: 16
	_delta: 8
	elementType (ptr): 0x8066598
	elementType[0-9]:
elementType[0]: ConstantElementType (ptr): 0x8064bf0
ElementType (ptr): 0x8064bf0
Stg_Class (ptr): 0x8064bf0
	sizeOfSelf: 48
	_deleteSelf: Yes
	type: ConstantElementType
	_delete (func ptr): 0x804be50
	_print (func ptr): 0x804be79
	_copy (func ptr): (nil)
	_build (func ptr): 0x804becb
	_evaluateShapeFunctionsAt (func ptr): 0x804bed0
	_evaluateShapeFunctionLocalDerivsAt (func ptr): 0x804beeb
	nodeCount: 1
elementType[1]: BilinearElementType (ptr): 0x8066360
ElementType (ptr): 0x8066360
Stg_Class (ptr): 0x8066360
	sizeOfSelf: 96
	_deleteSelf: Yes
	type: BilinearElementType
	_delete (func ptr): 0x804c128
	_print (func ptr): 0x804c151
	_copy (func ptr): (nil)
	_build (func ptr): 0x804c2f7
	_evaluateShapeFunctionsAt (func ptr): 0x804c2fc
	_evaluateShapeFunctionLocalDerivsAt (func ptr): 0x804c3b7
	nodeCount: 4
self->minElLocalCoord: (-1.00000,-1.00000,)
self->maxElLocalCoord: (1.00000,1.00000,)
self->elLocalLength: (2.00000,2.00000,)
elementType[2]: TrilinearElementType (ptr): 0x80663c8
ElementType (ptr): 0x80663c8
Stg_Class (ptr): 0x80663c8
	sizeOfSelf: 120
	_deleteSelf: Yes
	type: TrilinearElementType
	_delete (func ptr): 0x804c7b0
	_print (func ptr): 0x804c7d9
	_copy (func ptr): (nil)
	_build (func ptr): 0x804c82b
	_evaluateShapeFunctionsAt (func ptr): 0x804c830
	_evaluateShapeFunctionLocalDerivsAt (func ptr): 0x804c9ee
	nodeCount: 8
elementType[3]: ConstantElementType (ptr): 0x8066448
ElementType (ptr): 0x8066448
Stg_Class (ptr): 0x8066448
	sizeOfSelf: 48
	_deleteSelf: Yes
	type: TestElementType0
	_delete (func ptr): 0x804be50
	_print (func ptr): 0x804be79
	_copy (func ptr): (nil)
	_build (func ptr): 0x804becb
	_evaluateShapeFunctionsAt (func ptr): 0x804bed0
	_evaluateShapeFunctionLocalDerivsAt (func ptr): 0x804beeb
	nodeCount: 1
elementType[4]: ConstantElementType (ptr): 0x8066480
ElementType (ptr): 0x8066480
Stg_Class (ptr): 0x8066480
	sizeOfSelf: 48
	_deleteSelf: Yes
	type: TestElementType1
	_delete (func ptr): 0x804be50
	_print (func ptr): 0x804be79
	_copy (func ptr): (nil)
	_build (func ptr): 0x804becb
	_evaluateShapeFunctionsAt (func ptr): 0x804bed0
	_evaluateShapeFunctionLocalDerivsAt (func ptr): 0x804beeb
	nodeCount: 1
elementType[5]: ConstantElementType (ptr): 0x80664b8
ElementType (ptr): 0x80664b8
Stg_Class (ptr): 0x80664b8
	sizeOfSelf: 48
	_deleteSelf: Yes
	type: TestElementType2
	_delete (func ptr): 0x804be50
	_print (func ptr): 0x804be79
	_copy (func ptr): (nil)
	_build (func ptr): 0x804becb
	_evaluateShapeFunctionsAt (func ptr): 0x804bed0
	_evaluateShapeFunctionLocalDerivsAt (func ptr): 0x804beeb
	nodeCount: 1
elementType[6]: ConstantElementType (ptr): 0x80664f0
ElementType (ptr): 0x80664f0
Stg_Class (ptr): 0x80664f0
	sizeOfSelf: 48
	_deleteSelf: Yes
	type: TestElementType3
	_delete (func ptr): 0x804be50
	_print (func ptr): 0x804be79
	_copy (func ptr): (nil)
	_build (func ptr): 0x804becb
	_evaluateShapeFunctionsAt (func ptr): 0x804bed0
	_evaluateShapeFunctionLocalDerivsAt (func ptr): 0x804beeb
	nodeCount: 1
elementType[7]: ConstantElementType (ptr): 0x8066528
ElementType (ptr): 0x8066528
Stg_Class (ptr): 0x8066528
	sizeOfSelf: 48
	_deleteSelf: Yes
	type: TestElementType4
	_delete (func ptr): 0x804be50
	_print (func ptr): 0x804be79
	_copy (func ptr): (nil)
	_build (func ptr): 0x804becb
	_evaluateShapeFunctionsAt (func ptr): 0x804bed0
	_evaluateShapeFunctionLocalDerivsAt (func ptr): 0x804beeb
	nodeCount: 1
elementType[8]: ConstantElementType (ptr): 0x8066560
ElementType (ptr): 0x8066560
Stg_Class (ptr): 0x8066560
	sizeOfSelf: 48
	_deleteSelf: Yes
	type: TestElementType5
	_delete (func ptr): 0x804be50
	_print (func ptr): 0x804be79
	_copy (func ptr): (nil)
	_build (func ptr): 0x804becb
	_evaluateShapeFunctionsAt (func ptr): 0x804bed0
	_evaluateShapeFunctionLocalDerivsAt (func ptr): 0x804beeb
	nodeCount: 1
	]
